docs/guides/pallas_kernels_performance.md

```diff
@@ -0,0 +1,230 @@
+
+
+## **🧱 Core Pallas Design Patterns**
+
+These are the common techniques used in MaxText's Pallas kernels.
+
+* **Tiling & Blocking:** Move just a tile that fits on-chip, compute on it, and write it back.  
+* **Explicit Pipelining:** Overlap HBM↔VMEM loads with compute to hide latency (e.g., double-buffering).  
+* **Online Accumulation:** Combine partial results as you go; don’t materialize huge intermediate arrays.  
+* **Auxiliary Metadata:** Precompute control tables (e.g., token-to-expert ranges) and keep them in fast scalar memory.  
+* **Compute↔Communication Overlap:** In distributed runs, overlap local work with cross-device traffic when possible.
+
+
+## **✍️ Writing & Integrating a Kernel**
+
+A Pallas kernel is a plain Python function that operates on Refs (references to array tiles). You pass the function to pl.pallas\_call, which compiles and schedules it.
+
+### **Example 1: Minimal Elementwise Add**
+
+This is a simple example to show the basic structure.
+
+
+```python
+import jax  
+import jax.numpy as jnp  
+from jax.experimental import pallas as pl
+
+def add_vectors_kernel(x_ref, y_ref, o_ref):  
+  o_ref[:] = x_ref[:] + y_ref[:]
+
+def add_vectors(x: jax.Array, y: jax.Array) -> jax.Array:  
+  assert x.shape == y.shape  
+  return pl.pallas_call(  
+      add_vectors_kernel,  
+      out_shape=jax.ShapeDtypeStruct(shape=x.shape, dtype=x.dtype),  
+  )(x, y)
+```
+
+### **Example 2: Blocked 2D Add with BlockSpec**
+
+This example shows how to map a grid of blocks over larger arrays.
+
+```python
+import jax
+import jax.numpy as jnp
+from jax.experimental import pallas as pl
+
+def tile_add_kernel(x_ref, y_ref, o_ref):
+  # Operate on the tile slices handed in by BlockSpecs (already in VMEM on TPU).
+  o_ref[:, :] = x_ref[:, :] + y_ref[:, :]
+
+def tile_add(x: jax.Array, y: jax.Array) -> jax.Array:
+  assert x.shape == y.shape and x.ndim == 2
+  B0 = min(128, x.shape[0])  # Example choice; tune this with a sweep
+  B1 = x.shape[1]            # Full width tile (for illustration)
+
+  # Map program id (tile index) -> tile origin in the full (HBM) array.
+  # NOTE: The runtime advances origins by `block_shape`, so `i` is already a tile
+  # index. Do NOT multiply by B0 here.
+  in_out_spec = pl.BlockSpec(
+      block_shape=(B0, B1),
+      index_map=lambda i: (i, 0),
+  )
+
+  # Grid is implied by data & block shape (use ceiling-division helper).
+  grid = (pl.cdiv(x.shape[0], B0),)
+  # Note: grid size can also be computed dynamically at runtime.
+
+  return pl.pallas_call(
+      tile_add_kernel,
+      out_shape=jax.ShapeDtypeStruct(x.shape, x.dtype),
+      in_specs=[in_out_spec, in_out_spec],
+      out_specs=in_out_spec,
+      grid=grid,
+  )(x, y)
+```
+
+**Tip:** In practice, you’ll sweep (B0, B1). Focus tuning on block shapes; treat grid as derived.
```

Comment:
This may be jargon that I don't know, but can you explain what "sweep" means in this context?